src/server/common/server_base.cpp

```diff
@@ -128,19 +130,57 @@ void server_base::serve(boost::asio::yield_context yield, boost::asio::ip::tcp::
 
 	DBG_SERVER << client_address(socket) << "\tnew connection tentatively accepted\n";
 
-	boost::shared_array<char> handshake(new char[4]);
-	async_read(*socket, boost::asio::buffer(handshake.get(), 4), yield[error]);
+	union {
+		uint32_t number;
+		char buf[4];
+	} protocol_version;
+
+	union {
+		uint32_t number;
+		char buf[4];
+	} handshake_response;
+	
+	any_socket_ptr final_socket;
+
+	async_read(*socket, boost::asio::buffer(protocol_version.buf), yield[error]);
 	if(check_error(error, socket))
 		return;
 
-	if(memcmp(handshake.get(), "\0\0\0\0", 4) != 0) {
-		ERR_SERVER << client_address(socket) << "\tincorrect handshake\n";
-		return;
-	}
+	switch(ntohl(protocol_version.number)) {
```

Comment:
Only reason to use union here is to avoid warning from gcc about type-punning. It's consistent with other places in the code that use same trick and writing/reading to different union members is the core of it.

Reply:
https://en.cppreference.com/w/cpp/language/union#Explanation says: 

> It's undefined behavior to read from the member of the union that wasn't most recently written.

So unless all usable compilers feature a non-standard extension that makes this defined it doesn't seem like something we should do.

Reply:
My solution was influenced by https://stackoverflow.com/questions/98650/what-is-the-strict-aliasing-rule or some precursor of it. We might find another solution but it's out of scope for this PR.

Reply:
> https://en.cppreference.com/w/cpp/language/union#Explanation says:
> 
> > It's undefined behavior to read from the member of the union that wasn't most recently written.
> 
> So unless all usable compilers feature a non-standard extension that makes this defined it doesn't seem like something we should do.

When it comes to sockets-related things, that appears to essentially be the case.

Nevertheless, relying on union type-punning, host byte order and `ntohl`/`htonl` is silly when you can simply explicitly define the protocol byte order instead.

Reply:
And how exactly would such explicit definition be implemented? You mean comparing exact byte patterns instead of casting them to numbers?

Reply:
No, you do convert them to numbers, but without ntohl or type-punning.

Something like `uint32 addr = uint32(byte0) << 24 | uint32(byte1) << 16 | uint32(byte2) << 8 | uint32(byte3);` where each byte is a uint8, explicitly defines the conversion without any type-punning or byte order dependencies.

Wrap it and the inverse in a function, and you're done.

You can then do stuff like 

    read_into(buf)
    switch (big_endian_uint32(buf)) {

with hopefully better names than I came up with.

Reply:
That just seems more error prone to me. Sending a single number in network byte order is equally good protocol definition and bit sizes are fixed too due to use of appropriate types. htonl/ntohl calls are more maintainable than manual bit shifts.